### ISSUE TITLE
Reconfigure emulation flags

### ISSUE THREAD

**Issue by paulirish** [GoogleChrome/lighthouse#10910]
**_TLDR_**: The emulation flags are awkward and there's implicit assumptions made. We should fix that and support our varied usecases with first-class settings. Key idea: get rid of `emulatedFormFactor=none`.

---------

DevTools, Calibre, and WPT are three lighthouse clients that handle emulation previous to lighthouse running.  In DevTools' case, only screen emulation is applied, and lighthouse is expected to apply its own network emulation. And calibre's case, both emulations are applied and the intent is for a lighthouse to not apply double emulation anywhere. LH on WPT supports a few approaches and [everyone's been confused](https://github.com/WPO-Foundation/wptagent/pull/297) at some point.

We introduced the illustrious `internalDisableDeviceScreenEmulation` flag (#9377) to solve this. (However even with this, LH will still overwrite any previously set userAgentOverride). But if `emulatedFormFactor` (#6098) is `none` _and_ external UA emulation is applied, you don't get correct scoring.

This is messy and I think that there's an opportunity to completely redefine this configuration to be more clear. 

### Goals

1. no surprises wrt mobile or desktop scoring being used
1. a more first-class config for this external-emulation setup
1. continue to support the no-emulation-lh-is-running-against-a-real-phone scenario

### Starting Proposal

* Rename `emulatedFormFactor` to `formFactor`. This is the key flag and the value **must** be either `mobile` or `desktop`, `none` (née `provided`) is no longer a valid option. Our perf metrics and 3 SEO metrics change their behavior based on mobile v desktop, so it seems important to make this an explicit user contract.
* Add two more flags: `disableScreenEmulation` and `disableNetworkEmulation`.
* Remove `internalDisableDeviceScreenEmulation`. Throw if this or `emulatedFormFactor` are used.
* Internally, we drop the _guessed_ `TestedAsMobileDevice` bool and just use the `formFactor` enum of `'desktop'|'mobile'`.

The combination of these allows us to support the below user stories with clear expectations. It also prevents the case of a using incorrect scoring accidentally.

#### User stories

I believe we have 4 user stories to support (right?)

1. Typical LH (we apply emulation): `formFactor` is set.
1. DevTools: `formFactor` is set via the mobile/desktop radio, and `screenEmulation` is false.
1. Calibre and WPT:  `formFactor` is set, and both `screenEmulation: false, throttlingMethod: 'provided'`.  `emulatedUserAgent` is set as desired.
1. LH on mobile device: `--no-screenEmulation` and `--throttling.cpuSlowdownMultiplier=1`. (`--formFactor=mobile` is the default already)

### Extra thoughts and questions

* Given this situation and our different desktop scoring, it definitely raises the priority of communicating mobile/desktop run near the score.  #8178 #9379 #9438 
* Does the CLI default to `--formFactor=mobile` or do we make it (annoyingly) required? 
   - [the TestedAsMobileDevice logic](https://github.com/GoogleChrome/lighthouse/blob/7109852b/lighthouse-core/gather/gather-runner.js#L482-L486) correctly handles real-mobile-device, but doesn't handle external UA emulation. We _could_ detect there's external UA emulation being applied and throw if we don't see configuration that matches. Or with the proposal implemented, we may just log a warning to let them know we see it. 
   - [docs: Testing on a mobile device](https://github.com/GoogleChrome/lighthouse/blob/master/docs/readme.md#testing-on-a-mobile-device) still works correctly in v6, works in v6 even with `emulated-form-factor=none`. We could also detect a potential double-emulation scenario here and warn to make sure flags are correctly setup.
* https://github.com/GoogleChrome/lighthouse/pull/9377#pullrequestreview-264464259 has some discussion around custom UA/metrics/touch/etc. I think the above proposal lays good groundwork for allowing that configurability.
* Throttling is related, but I haven't yet considered if we should also reconfigure those settings to better align with these. Probably not, but we could optionally detect and warn on unexpected emulation/throttling pairings.
   - Desktop throttling preset.. Going through all this I see that #10322 made it so devtools/desktop use the same throttling constants as lr-desktop, but CLI/desktop doesn't. It probly should? Will file as a separate issue.
* I was a lil concerned the LH v6 on WPT might have broke something.  But looks like https://github.com/WPO-Foundation/wptagent/pull/297 which was merged last week fixes things (judging by [Joseph's last comment](https://github.com/WPO-Foundation/wptagent/pull/297#issuecomment-635004026)) - _Update_: [there's a bug](https://github.com/paulirish/lh-scorecalc/pull/16#issuecomment-639146430) but only for the scorecalc handoff, actual scoring is all good. 

----------

**References**

- Probably related emulation-y issues: #10836 https://github.com/GoogleChrome/lighthouse-ci/issues/321 https://github.com/WPO-Foundation/wptagent/pull/297
- lol: #7044 
- Mobile run (with external screen & UA emulation), scored as desktop: https://googlechrome.github.io/lighthouse/viewer/?gist=701d46b57b8252011a25025b2442ca81

**Comment by paulirish** [comment]
cc @wildlyinaccurate I have a feeling you'd have been much happier with the above proposal, so you wouldn't have had to do all the investigation in https://github.com/WPO-Foundation/wptagent/pull/297. But curious for your thoughts. :)

**Comment by connorjclark** [comment]
Related: I think we may need something like `setEmulation` in the Lighthouse runner API so that we can reset the emulation correctly for programmatic usages. See https://github.com/GoogleChrome/lighthouse/pull/10716#discussion_r428943503

**Comment by patrickhulce** [comment]
> Rename emulatedFormFactor to formFactor. This is the key flag and the value must be either mobile or desktop, none (née provided) is no longer a valid option.

Sounds great, love it 👍 

> Add disableScreenEmulation

Basically required to do the above, SGTM 👍 

> Remove internalDisableDeviceScreenEmulation. Throw if this or emulatedFormFactor are used.

Sounds like good housekeeping given the above 👍 

> Internally, we drop the guessed TestedAsMobileDevice bool and just use the formFactor enum of 'desktop'|'mobile'.

Feels a little weird that `formFactor` is at this point just acting more like `scoringMode: 'easy' | 'hard'` instead of what the form factor actually was. Not sure I completely buy into dropping `TestedAsMobileDevice` for double checking it matched `formFactor`

> Or with the proposal implemented, we may just log a warning to let them know we see it.

Speak of the 👿  Yeah let's do that instead :)

> Add disableNetworkEmulation

This is the one I don't understand :) Sounds like it's just returning us to the old days of duplicate flags controlling the same thing.

**Comment by brendankenny** [comment]
Old comment from 

> If we were designing our flags from scratch we definitely wouldn't want to overlap with emulatedFormFactor like this :/
>
> The combo with provided makes emulatedFormFactor actually emulatedUAString, and if you're just trying to get your settings right, it's not clear why you'd want emulatedViewportMethod vs emulatedFormFactor and what each one does (and is Lighthouse providing provided or am I?).
>
> See below for one suggestion to simplify this, but what if we just went all the way and added free-form emulation variables for UA string, metrics, and touchEnabled (in the style of throttling)? emulatedFormFactor would stay a set of presets that most would use, but it would be possible to override. This would also unlock what PaulKinlan is looking for re: feature phones.

Not sure if I completely agree with this, but just splitting up all our emulation options so that they don't have interactions (unexpected or otherwise) anymore does have a certain appeal. Most people will still use the presets, and the people heavily tweaking will want to heavily tweak anyways.

**Comment by paulirish** [comment]
discussed. plan:

1. lets expose screen emu properties (width, height, dpr) to config. 
1. let's expose UA to config.
1. determine disable* in relation to these.
1. perhaps in the calibre/wpt "provided" throttling setup, we should have more useful text in runtime settings. do they want to set a string or .... ?
![image](https://user-images.githubusercontent.com/39191/84708834-0f5b3b80-af16-11ea-8a87-a90ffe5512d2.png)

**Comment by benschwarz** [comment]
Thanks for the ping @paulirish ✌️

> lets expose screen emu properties (width, height, dpr) to config.
> let's expose UA to config.
> determine disable* in relation to these.

SGTM. 

> perhaps in the calibre/wpt "provided" throttling setup, we should have more useful text in runtime settings. do they want to set a string or .... ?

Yeah, for completeness I think it'd be good to be able to set some properties that explain the test environment more deeply. I raised this on #7053 but never took it anywhere.

**Comment by paulirish** [comment]
oh word! k we'll break off my number 4 into #7053. perfect.

**Comment by amannn** [comment]
Thanks for your great work on lighthouse!

I've experienced some different scores when running lighthouse on CI than when running locally in regards to the total blocking time.

After some research, I think this is due to the `cpuSlowdownMultiplier` option as it's relative to the available CPU power. My CI runner already is a bit less capable, therefore I was seeing lower numbers there than when testing locally on my MacBook Pro.

I was wondering if there's a way to normalize this behaviour across environments, so the score would be the same, regardless of the machine? Obviously, there would be some minimum CPU requirements, but ideally from there on the result would be the same. I'm not sure if this is somehow possible practically?

**Comment by patrickhulce** [comment]
I think you'd be interested in the saga of https://github.com/GoogleChrome/lighthouse/issues/9085 and the resulting [calibration documentation](https://github.com/GoogleChrome/lighthouse/blob/fe74339f1ee2b5606d5a7af19b6b52bd622e862a/docs/throttling.md#calibrating-multipliers) @amannn :)

tl;dr - we tried to do this, but it's very, very difficult and even benchmarks built by *dedicated benchmark companies* can't predict machine performance well enough to normalize Lighthouse scores, so a single contributor here using 20% of their time won't be able to solve it either :/

**Comment by amannn** [comment]
@patrickhulce Oh right, that calibration guide is really helpful! I definitely understand that this is a hard problem.

Thank you for your help!
